internal/dockercompose/client.go

```diff
@@ -136,18 +142,27 @@ func (c *cmdDCClient) StreamLogs(ctx context.Context, configPaths []string, serv
 
 	err = cmd.Start()
 	if err != nil {
-		return nil, errors.Wrapf(err, "`docker-compose %s`",
+		return nil, errors.Wrapf(err, "cmd `docker-compose %s` failed to start",
 			strings.Join(args, " "))
 	}
 
+	// extra pipe is created so that we can ensure we've read all output from stdout _before_ calling Cmd::Wait()
+	// AND so that we can propagate errors (vs. just silently logging them) to the reader so they can choose to
+	// restart on failures
+	r, w := io.Pipe()
 	go func() {
-		err = cmd.Wait()
-		if err != nil {
-			logger.Get(ctx).Debugf("cmd `docker-compose %s` exited with error: \"%v\" (stderr: %s)",
-				strings.Join(args, " "), err, errBuf.String())
+		// prefer reporting process errors to copy errors as they're more likely to be the root cause
+		_, copyErr := io.Copy(w, stdout)
```

Comment:
this seems very strange.

I've read a ton of code that uses the Go os/exec standard lib. 

I've never seen code that used it this way, in particular, blocking a cmd.Wait() on an io.Copy.

Can you talk a little bit more about the conditions where the copy failed but the command succeeded? That would be very surprising to me, based on my understanding of how Cmd/Stdout is implemented internally.

i think i would have expected something more like:
```
r, w := io.Pipe()
cmd.Stdout = w
go func() {
		err = cmd.Wait()
...
}()
return r, nil
```

and not use the built-in StdoutPipe() at all

Reply:
Agh, yeah, I originally had something similar to that and then was having issues with something unrelated and eventually ended up in this weird hybrid state 😧

Massively simplified in 058b0561b.